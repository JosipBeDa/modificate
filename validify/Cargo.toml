[package]
authors = ["biblius <crazycompanyxxl@gmail.com>"]
description = "An extension of the validator crate that enables payload modification"
edition = "2021"
exclude = ["derive_tests/"]
homepage = "https://github.com/biblius/validify"
keywords = ["modify", "payload", "validate", "validify"]
license = "MIT"
name = "validify"
readme = "../README.md"
repository = "https://github.com/biblius/validify"
version = "0.1.0"


[dependencies]
card-validate = { version = "2.3" }
idna = "0.3"
indexmap = { version = "1", features = ["serde-1"] }
lazy_static = "1"
phonenumber = { version = "0.3.2" }
proc-macro2 = "1.0.47"
regex = "1"
serde = "1.0.152"
serde_derive = "1"
serde_json = "1"
syn = { version = "1.0.103", features = ["full"] }
unic-ucd-common = { version = "0.9" }
url = "2"
validify_derive = { version = "0.1.0", path = "../validify_derive" }
