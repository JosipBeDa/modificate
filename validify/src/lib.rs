//! A procedural macro built on top of the [validator](https://docs.rs/validator/latest/validator/)
//! crate that provides attributes for field modifiers. Particularly useful in the context of web payloads.
//!
//! Visit the [repository](https://github.com/biblius/validify) to see exactly how it works.
//!
//!  ### Example
//!
//! ```
//! use validify::Validify;
//!
//! #[derive(Debug, Clone, serde::Serialize, serde::Deserialize, Validify)]
//! struct Testor {
//!     #[modify(lowercase, trim)]
//!     #[validate(length(equal = 8))]
//!     pub a: String,
//!     #[modify(trim, uppercase)]
//!     pub b: Option<String>,
//!     #[modify(custom(do_something))]
//!     pub c: String,
//!     #[modify(custom(do_something))]
//!     pub d: Option<String>,
//!     #[validify]
//!     pub nested: Nestor,
//! }
//!
//! #[derive(Debug, Clone, serde::Serialize, serde::Deserialize, Validify)]
//! struct Nestor {
//!     #[modify(trim, uppercase)]
//!     #[validate(length(equal = 12))]
//!     a: String,
//!     #[modify(capitalize)]
//!     #[validate(length(equal = 14))]
//!     b: String,
//! }
//!
//! fn do_something(input: &mut String) {
//!     *input = String::from("modified");
//! }
//!
//! let mut test = Testor {
//!   a: "   LOWER ME     ".to_string(),
//!   b: Some("  makemeshout   ".to_string()),
//!   c: "I'll never be the same".to_string(),
//!   d: Some("Me neither".to_string()),
//!   nested: Nestor {
//!     a: "   notsotinynow   ".to_string(),
//!       b: "capitalize me.".to_string(),
//!   },
//! };
//!
//! // The magic line
//! let res = Testor::validify(test.into());
//!
//! assert!(matches!(res, Ok(_)));
//!
//! let test = res.unwrap();
//!
//! // Parent
//! assert_eq!(test.a, "lower me");
//! assert_eq!(test.b, Some("MAKEMESHOUT".to_string()));
//! assert_eq!(test.c, "modified");
//! assert_eq!(test.d, Some("modified".to_string()));
//! // Nested
//! assert_eq!(test.nested.a, "NOTSOTINYNOW");
//! assert_eq!(test.nested.b, "Capitalize me.");
//! ```

mod error;
mod traits;
mod validation;

pub use validation::cards::validate_credit_card;
pub use validation::contains::validate_contains;
pub use validation::email::validate_email;
pub use validation::ip::{validate_ip, validate_ip_v4, validate_ip_v6};
pub use validation::length::validate_length;
pub use validation::must_match::validate_must_match;
pub use validation::non_control_char::validate_non_control_character;
pub use validation::phone::validate_phone;
pub use validation::r#in::validate_in;
pub use validation::range::validate_range;
pub use validation::required::validate_required;
pub use validation::urls::validate_url;

pub use error::{ValidationError, ValidationErrors};
pub use validify_derive::{schema_validation, Validate, Validify};

/// Validates structs based on the provided `validate` parameters. Can be implemented on its own if one doesn't need payload modifications.
pub trait Validate {
    fn validate(&self) -> Result<(), ValidationErrors>;
}

impl<T: Validate> Validate for &T {
    fn validate(&self) -> Result<(), ValidationErrors> {
        T::validate(*self)
    }
}

/// Modifies struct based on the provided `modify` parameters. Automatically implemented when deriving Validify.
pub trait Modify {
    /// Apply the provided modifiers to self
    fn modify(&mut self);
}

/// Combines `Validate` and `Modify` in one trait and provides the intermediary payload struct. This trait should never be implemented manually,
/// and should be derived with the `#[validify]` macro which automatically implements `Validate`, `Modify` and `Validify`.
pub trait Validify: Modify + Validate + Sized + From<Self::Payload> {
    /// Represents the same structure of the implementing struct,
    /// with all its fields as options. Used to represent a completely
    /// deserializable version of the struct even if the fields are missing.
    /// Used for more detailed descriptions of what fields are missing, along
    /// with any other validation errors.
    ///
    /// This type is automatically implemented when deriving validify by creating
    /// an accompanying payload struct:
    ///
    /// ```ignore
    /// #[validify]
    /// #[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
    /// struct Data {
    ///     a: String
    /// }
    ///
    /// // expands to
    /// struct DataPayload {
    ///     a: Option<String>
    /// }
    ///
    /// /*
    ///  * serde impls and other stuff
    ///  */
    ///
    /// impl Validify for Data {
    ///     type Payload = DataPayload;
    ///
    ///     /* fn validify(payload: Self::Payload) { ... } */
    /// }
    ///
    /// ```
    type Payload: serde::de::DeserializeOwned + Validate;

    /// Apply the provided modifiers to self and run validations.
    fn validify(payload: Self::Payload) -> Result<Self, ValidationErrors> {
        // Since the payload is all options, this will
        // only check if there are missing required fields
        <Self::Payload as Validate>::validate(&payload)?;
        let mut this = Self::from(payload);
        <Self as Modify>::modify(&mut this);
        <Self as Validate>::validate(&this)?;
        Ok(this)
    }
}

#[macro_export]
/// Designed to be used with the [schema_validation] proc macro. Used for ergonomic custom error handling.
///
/// Adds a schema validaton error to the generated `ValidationErrors`.
///
/// The errors argument should pass in an instance of `ValidationErrors`,
/// and usually is used with the one generated from `schema_validation`.
///
/// Accepts:
///
/// `("code", errors)`
/// `("code", "custom message", errors)`
macro_rules! schema_err {
    ($code:literal, $errors:expr) => {
        $errors.add(::validify::ValidationError::new_schema($code));
    };
    ($code:literal, $message:literal, $errors:expr) => {
        $errors
            .add(::validify::ValidationError::new_schema($code).with_message($message.to_string()));
    };
    ($code:literal, $message:expr, $errors:expr) => {
        $errors.add(::validify::ValidationError::new_schema($code).with_message($message));
    };
}

#[macro_export]
/// Designed to be used with the [schema_validation] proc macro. Used for ergonomic custom error handling.
///
/// Adds a field validaton error to the generated `ValidationErrors`
///
///  The errors argument should pass in an instance of `ValidationErrors`,
/// and usually is used with the one generated from `schema_validation`.
///
/// Accepts:
///
/// `("field_name", "code", errors)`
/// `("field_name", "code", "custom message", errors)`
macro_rules! field_err {
    ($field:literal, $code:literal, $errors:expr) => {
        $errors.add(::validify::ValidationError::new_field($field, $code));
    };
    ($field:literal, $code:literal, $message:literal, $errors:expr) => {
        $errors.add(
            ::validify::ValidationError::new_field($field, $code)
                .with_message($message.to_string()),
        );
    };
    ($field:literal, $code:literal, $message:expr, $errors:expr) => {
        $errors.add(::validify::ValidationError::new_field($field, $code).with_message($message));
    };
}
