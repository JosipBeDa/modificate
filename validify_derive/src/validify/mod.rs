pub mod r#impl;
pub mod quoting;
