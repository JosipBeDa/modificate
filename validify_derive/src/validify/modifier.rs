use crate::fields::FieldInfo;
use quote::quote;

#[derive(Debug, PartialEq, Eq)]
pub enum Modifier {
    Trim,
    Uppercase,
    Lowercase,
    Capitalize,
    Custom { function: syn::Path },
    Nested,
}

impl Modifier {
    /// Returns direct modification tokens as the first element and any nested validify tokens as the second element.
    /// Necessary because we need both in case a nested validify occurs. In that case, the first element will have the
    /// necessary modification tokens for nested elements in the `Modify` impl while the second will have the tokens
    /// for the `Validify` impl.
    pub fn to_validify_tokens(&self, field_info: &FieldInfo) -> proc_macro2::TokenStream {
        let param = field_info.quote_modifier_param();

        match self {
            Modifier::Trim => {
                let tokens = if field_info.ident_override.is_some() || field_info.is_option() {
                    quote!(
                        *#param = #param.trim().to_string();
                    )
                } else {
                    quote!(
                        #param = #param.trim().to_string();
                    )
                };
                field_info.wrap_modifier_if_option(
                    field_info.wrap_modifier_if_collection(param, tokens, self),
                )
            }
            Modifier::Uppercase => {
                let tokens = if field_info.ident_override.is_some() || field_info.is_option() {
                    quote!(
                        *#param = #param.to_uppercase();
                    )
                } else {
                    quote!(
                        #param = #param.to_uppercase();
                    )
                };
                field_info.wrap_modifier_if_option(
                    field_info.wrap_modifier_if_collection(param, tokens, self),
                )
            }
            Modifier::Lowercase => {
                let tokens = if field_info.ident_override.is_some() || field_info.is_option() {
                    quote!(
                        *#param = #param.to_lowercase();
                    )
                } else {
                    quote!(
                        #param = #param.to_lowercase();
                    )
                };
                field_info.wrap_modifier_if_option(
                    field_info.wrap_modifier_if_collection(param, tokens, self),
                )
            }
            Modifier::Capitalize => {
                let tokens = if field_info.ident_override.is_some() || field_info.is_option() {
                    quote!(
                      *#param = ::std::format!("{}{}", &#param[0..1].to_uppercase(), &#param[1..]);
                    )
                } else {
                    quote!(
                      #param = ::std::format!("{}{}", &#param[0..1].to_uppercase(), &#param[1..]);
                    )
                };
                field_info.wrap_modifier_if_option(
                    field_info.wrap_modifier_if_collection(param, tokens, self),
                )
            }
            Modifier::Custom { function } => {
                let tokens = if field_info.ident_override.is_some() || field_info.is_option() {
                    quote!(
                        #function(#param);
                    )
                } else {
                    quote!(
                        #function(&mut #param);
                    )
                };
                field_info.wrap_modifier_if_option(tokens)
            }
            Modifier::Nested => {
                let modifications = if field_info.is_list() {
                    quote!(
                        for el in #param.iter_mut() {
                            el.modify();
                        }
                    )
                } else {
                    quote!(#param.modify();)
                };

                field_info.wrap_modifier_if_option(modifications)
            }
        }
    }
}
