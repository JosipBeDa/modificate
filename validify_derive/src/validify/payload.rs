use crate::fields::FieldInfo;
use proc_macro2::{Ident, Span};
use proc_macro_error::abort;
use quote::quote;

use syn::spanned::Spanned;

pub(super) fn generate(input: &syn::DeriveInput) -> (proc_macro2::TokenStream, syn::Ident) {
    let ident = &input.ident;
    let attributes = input
        .attrs
        .iter()
        .filter(|attr| attr.meta.path().is_ident("serde"))
        .collect::<Vec<_>>();
    let visibility = &input.vis;

    let payload_ident = syn::Ident::new(
        &format!("{}Payload", &input.ident.to_string()),
        Span::call_site(),
    );

    let fields = FieldInfo::collect(input);

    let payload_fields = fields
        .iter()
        .map(map_payload_fields)
        .collect::<Vec<proc_macro2::TokenStream>>();

    let (impl_generics, ty_generics, where_clause) = input.generics.split_for_impl();

    let into_fields = fields
        .iter()
        .map(map_into_fields)
        .collect::<Vec<proc_macro2::TokenStream>>();

    let from_fields = fields
        .iter()
        .map(map_from_fields)
        .collect::<Vec<proc_macro2::TokenStream>>();

    let quoted = quote!(
        #[derive(Debug, Clone, ::validify::Validate, serde::Deserialize)]
        #(#attributes)*
        #visibility struct #payload_ident #ty_generics #where_clause {
            #(#payload_fields)*
        }

        impl #impl_generics From<#ident> for #payload_ident {
            fn from(original: #ident) -> Self {
                Self {
                    #(#into_fields)*
                }
            }
        }

        impl #impl_generics From<#payload_ident> for #ident {
            fn from(payload: #payload_ident) -> Self {
                Self {
                    #(#from_fields)*
                }
            }
        }
    );

    (quoted, payload_ident)
}

fn map_payload_fields(info: &FieldInfo) -> proc_macro2::TokenStream {
    let ident = info.field.ident.as_ref().unwrap();
    let is_list = info.is_list();
    let ty = &info.field.ty;

    if !info.is_option() {
        return payload_path(info);
    }

    if !info.is_nested_validify() {
        return quote!(#ident: #ty,);
    }

    let syn::Type::Path(mut path) = ty.clone() else {
        abort!(
            info.field.span(),
            "Nested validifes must be structs implementing Validify"
        )
    };

    let syn::PathArguments::AngleBracketed(ref mut args) =
        path.path.segments.last_mut().unwrap().arguments
    else {
        abort!(path.span(), "Cannot apply payload type to field")
    };

    let syn::GenericArgument::Type(syn::Type::Path(ref mut inner_path)) =
        args.args.last_mut().unwrap()
    else {
        abort!(path.span(), "Cannot apply payload type to field")
    };

    if is_list {
        payload_path_angle_bracketed(inner_path);
    } else {
        let type_ident = inner_path.path.segments.last().unwrap().ident.to_string();
        inner_path.path.segments.last_mut().unwrap().ident =
            Ident::new(&format!("{type_ident}Payload"), Span::call_site());
    }

    let payload_type = syn::Type::Path(path);

    quote!(
        #[validate]
        #ident: #payload_type,
    )
}

fn map_from_fields(info: &FieldInfo) -> proc_macro2::TokenStream {
    let ident = info.field.ident.as_ref().unwrap();

    if info.is_option() {
        if info.is_nested_validify() && info.is_list() {
            return quote!(
                #ident: payload.#ident.map(|v|v.into_iter().map(|el|el.into()).collect()),
            );
        }
        if info.is_nested_validify() {
            return quote!(
                #ident: payload.#ident.map(|o|o.into()),
            );
        }

        quote!(
            #ident: payload.#ident,
        )
    } else {
        if info.is_nested_validify() && info.is_list() {
            return quote!(#ident: payload.#ident.unwrap().into_iter().map(|el|el.into()).collect(),);
        }
        if info.is_nested_validify() {
            return quote!(#ident: payload.#ident.unwrap().into(),);
        }
        quote!(#ident: payload.#ident.unwrap(),)
    }
}

fn map_into_fields(info: &FieldInfo) -> proc_macro2::TokenStream {
    let ident = info.field.ident.as_ref().unwrap();

    if info.is_option() {
        if info.is_nested_validify() && info.is_list() {
            return quote!(#ident: original.#ident.map(|v| v.into_iter().map(|el|el.into()).collect()),);
        }

        if info.is_nested_validify() {
            return quote!(#ident: original.#ident.map(|o|o.into()),);
        }

        quote!(#ident: original.#ident,)
    } else {
        if info.is_nested_validify() && info.is_list() {
            return quote!(#ident: Some(original.#ident.into_iter().map(|el|el.into()).collect()),);
        }

        if info.is_nested_validify() {
            return quote!(#ident: Some(original.#ident.into()),);
        }

        quote!(#ident: Some(original.#ident),)
    }
}

fn payload_path_angle_bracketed(path: &mut syn::TypePath) {
    // Type is contained in a List<T>. It will always have angle args abd will
    // always be the last segment of the path
    let syn::PathArguments::AngleBracketed(ref mut args) =
        path.path.segments.last_mut().unwrap().arguments
    else {
        abort!(path.span(), "Cannot apply payload type to field")
    };

    let syn::GenericArgument::Type(syn::Type::Path(ref mut p)) = args.args.last_mut().unwrap()
    else {
        abort!(path.span(), "Cannot apply payload type to field")
    };

    let Some(segment) = p.path.segments.last_mut() else {
        abort!(p.span(), "Invalid path provided for validify")
    };

    segment.ident = Ident::new(&format!("{}Payload", segment.ident), Span::call_site());
}

fn payload_path(info: &FieldInfo) -> proc_macro2::TokenStream {
    let ident = &info.field.ident;
    let ty = &info.field.ty;

    if !info.is_nested_validify() {
        return quote!(
            #[validate(required)]
            #ident: Option<#ty>,
        );
    }

    let syn::Type::Path(mut path) = ty.clone() else {
        abort!(
            info.field.span(),
            "Nested validifes must be structs implementing Validify or collections/options of"
        )
    };

    if info.is_list() {
        payload_path_angle_bracketed(&mut path);
    } else {
        let ty_ident = path.path.segments.last().unwrap().ident.to_string();
        path.path.segments.last_mut().unwrap().ident =
            Ident::new(&format!("{ty_ident}Payload"), Span::call_site());
    }

    let payload_type = syn::Type::Path(path);

    quote!(
        #[validate(required)]
        #[validate]
        #ident: Option<#payload_type>,
    )
}

#[allow(dead_code)]
/// Could come in handy, parses the inner contents of an angle bracketed path and outputs
/// the original and the payload paths in a tuple
fn get_inner_path(ty: syn::Type) -> (syn::TypePath, syn::TypePath) {
    let syn::Type::Path(mut path) = ty else {
        abort!(
            ty.span(),
            "Nested validifes must be structs implementing Validify"
        )
    };
    let syn::PathArguments::AngleBracketed(ref mut args) =
        path.path.segments.last_mut().unwrap().arguments
    else {
        abort!(path.span(), "Cannot apply payload type to field")
    };
    let syn::GenericArgument::Type(syn::Type::Path(ref mut inner_path)) =
        args.args.last_mut().unwrap()
    else {
        abort!(path.span(), "Cannot apply payload type to field")
    };

    let original = inner_path.clone();

    let type_ident = inner_path
        .path
        .segments
        .last_mut()
        .unwrap()
        .ident
        .to_string();
    inner_path.path.segments.last_mut().unwrap().ident =
        Ident::new(&format!("{type_ident}Payload"), Span::call_site());

    (original, inner_path.clone())
}
