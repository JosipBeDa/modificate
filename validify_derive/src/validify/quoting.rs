use crate::{asserts::is_list, fields::FieldInformation, quoter::FieldQuoter, types::Modifier};
use quote::quote;

/// Creates a token stream applying the modifiers based on the field annotations.
pub(super) fn quote_field_modifiers(
    mut fields: Vec<FieldInformation>,
) -> (Vec<proc_macro2::TokenStream>, Vec<proc_macro2::TokenStream>) {
    let mut modifications = vec![];
    let mut nested_validifies = vec![];

    fields.drain(..).for_each(
        |FieldInformation {
             field,
             field_type,
             name,
             original_name,
             modifiers,
             ..
         }| {
            let field_ident = field.ident.unwrap();
            let field_quoter = FieldQuoter::new(field_ident, name, original_name, field_type);

            for modifier in modifiers.iter() {
                let (mods, nested) = quote_modifiers(&field_quoter, modifier);

                modifications.push(mods);

                if let Some(nested_validify) = nested {
                    nested_validifies.push(nested_validify)
                }
            }
        },
    );

    (modifications, nested_validifies)
}

/// Returns a modification and a validation (if it's nested) statement for the field.
fn quote_modifiers(
    field_quoter: &FieldQuoter,
    mod_type: &Modifier,
) -> (proc_macro2::TokenStream, Option<proc_macro2::TokenStream>) {
    let ty = field_quoter._type.clone();
    let modifier_param = field_quoter.quote_modifier_param();
    let is_option = field_quoter.check_option();
    let is_list = field_quoter.check_vec();

    let (mods, valids) = match mod_type {
        Modifier::Trim => (
            quote_trim_modifier(modifier_param, is_option, is_list),
            None,
        ),
        Modifier::Uppercase => (
            quote_uppercase_modifier(modifier_param, is_option, is_list),
            None,
        ),
        Modifier::Lowercase => (
            quote_lowercase_modifier(modifier_param, is_option, is_list),
            None,
        ),
        Modifier::Capitalize => (
            quote_capitalize_modifier(modifier_param, is_option, is_list),
            None,
        ),
        Modifier::Custom { function } => (
            quote_custom_modifier(modifier_param, function, is_option),
            None,
        ),
        Modifier::Nested => {
            let (modify, validate) = quote_nested_modifier(modifier_param, ty, is_option);
            (modify, Some(validate))
        }
    };

    (
        field_quoter.wrap_modifier_if_option(mods, false),
        valids.map(|tokens| field_quoter.wrap_modifier_if_option(tokens, true)),
    )
}

fn quote_nested_modifier(
    param: proc_macro2::TokenStream,
    ty: String,
    is_option: bool,
) -> (proc_macro2::TokenStream, proc_macro2::TokenStream) {
    let par = param.to_string();
    let field = par.split('.').last().unwrap();
    let field_ident: proc_macro2::TokenStream = format!("this.{field}").parse().unwrap();

    let is_list = is_list(&ty);

    let modifications = if is_list {
        quote!(
            for el in #param.iter_mut() {
                el.modify();
            }
        )
    } else {
        quote!(#param.modify();)
    };

    let param = is_option
        .then(|| {
            let field: proc_macro2::TokenStream = field.parse().unwrap();
            quote!(#field)
        })
        .unwrap_or(quote!(#field_ident));

    let nested_validifies = if is_list {
        quote!(
            for (i, el) in #param.iter_mut().enumerate() {
                if let Err(mut errs) = el.validify_self() {
                    errs.errors_mut().iter_mut().for_each(|err|err.set_location_idx(i, #field));
                    errors.merge(errs);
                }
            }
        )
    } else {
        quote!(
            if let Err(mut err) = #param.validify_self() {
                err.errors_mut().iter_mut().for_each(|e| e.set_location(#field));
                errors.merge(err);
            }
        )
    };

    (modifications, nested_validifies)
}

fn quote_custom_modifier(
    param: proc_macro2::TokenStream,
    function: &syn::Path,
    is_option: bool,
) -> proc_macro2::TokenStream {
    if is_option {
        quote!(
            #function(#param);
        )
    } else {
        quote!(
            #function(&mut #param);
        )
    }
}

pub(super) fn quote_trim_modifier(
    param: proc_macro2::TokenStream,
    is_option: bool,
    is_list: bool,
) -> proc_macro2::TokenStream {
    if is_list {
        return quote_vec_modifier(param, is_option, Modifier::Trim);
    }
    if is_option {
        quote!(
            *#param = #param.trim().to_string();
        )
    } else {
        quote!(
            #param = #param.trim().to_string();
        )
    }
}

pub(super) fn quote_uppercase_modifier(
    param: proc_macro2::TokenStream,
    is_option: bool,
    is_list: bool,
) -> proc_macro2::TokenStream {
    if is_list {
        return quote_vec_modifier(param, is_option, Modifier::Uppercase);
    }
    if is_option {
        quote!(
            *#param = #param.to_uppercase();
        )
    } else {
        quote!(
            #param = #param.to_uppercase();
        )
    }
}

fn quote_vec_modifier(
    param: proc_macro2::TokenStream,
    is_option: bool,
    mod_type: Modifier,
) -> proc_macro2::TokenStream {
    let modifier = match mod_type {
        Modifier::Trim => quote!(trim().to_string()),
        Modifier::Uppercase => quote!(to_uppercase()),
        Modifier::Lowercase => quote!(to_lowercase()),
        Modifier::Capitalize => quote!(),
        _ => unreachable!("Use of modifier that can be applied directly to vec forbidden"),
    };
    if is_option {
        if mod_type == Modifier::Capitalize {
            quote!(
                for el in #param.iter_mut() {
                    *el = ::std::format!("{}{}", &el[0..1].to_uppercase(), &el[1..])
                }
            )
        } else {
            quote!(
                for el in #param.iter_mut() {
                    *el = el.#modifier
                }
            )
        }
    } else if mod_type == Modifier::Capitalize {
        quote!(
            for el in #param.iter_mut() {
                *el = ::std::format!("{}{}", &el[0..1].to_uppercase(), &el[1..])
            }
        )
    } else {
        quote!(
            for el in #param.iter_mut() {
                *el = el.#modifier
            }
        )
    }
}

pub(super) fn quote_lowercase_modifier(
    param: proc_macro2::TokenStream,
    is_option: bool,
    is_list: bool,
) -> proc_macro2::TokenStream {
    if is_list {
        return quote_vec_modifier(param, is_option, Modifier::Lowercase);
    }
    if is_option {
        quote!(
            *#param = #param.to_lowercase();
        )
    } else {
        quote!(
            #param = #param.to_lowercase();
        )
    }
}

pub(super) fn quote_capitalize_modifier(
    param: proc_macro2::TokenStream,
    is_option: bool,
    is_list: bool,
) -> proc_macro2::TokenStream {
    if is_list {
        return quote_vec_modifier(param, is_option, Modifier::Capitalize);
    }
    if is_option {
        quote!(
          *#param = ::std::format!("{}{}", &#param[0..1].to_uppercase(), &#param[1..]);
        )
    } else {
        quote!(
          #param = ::std::format!("{}{}", &#param[0..1].to_uppercase(), &#param[1..]);
        )
    }
}
