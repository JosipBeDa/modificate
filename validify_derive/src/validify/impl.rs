use super::modifier::Modifier;
use crate::tokens::quote_field_modifiers;
use crate::{fields::FieldInfo, validate::r#impl::impl_validate};
use proc_macro_error::abort;
use quote::quote;
use syn::parenthesized;

const TRIM_MODIFIER: &str = "trim";
const CUSTOM_MODIFIER: &str = "custom";
const UPPERCASE_MODIFIER: &str = "uppercase";
const LOWERCASE_MODIFIER: &str = "lowercase";
const CAPITALIZE_MODIFIER: &str = "capitalize";
const VALIDIFY: &str = "validify";
const MODIFY: &str = "modify";

/// Impl entry point
pub fn impl_validify(input: &syn::DeriveInput) -> proc_macro2::TokenStream {
    let ident = &input.ident;

    let field_info = FieldInfo::collect(input);

    let (modifiers, nested_validifies) = quote_field_modifiers(field_info);

    let validate_impl = impl_validate(input);

    let (impl_generics, ty_generics, where_clause) = input.generics.split_for_impl();

    quote!(

    #validate_impl

    impl #impl_generics ::validify::Modify for #ident #ty_generics #where_clause {
        fn modify(&mut self) {
            #(#modifiers)*
        }
    }

    impl #impl_generics ::validify::Validify for #ident #ty_generics #where_clause {
        fn validify(&mut self) -> Result<(), ::validify::ValidationErrors> {
            let mut errors = ::validify::ValidationErrors::new();

            #(#nested_validifies)*

            <Self as ::validify::Modify>::modify(self);

            if let Err(errs) = <Self as ::validify::Validate>::validate(self) {
                errors.merge(errs);
            }

            if !errors.is_empty() {
                Err(errors)
            } else {
                Ok(())
            }
        }
    })
}

pub fn collect_modifiers(field: &syn::Field) -> Vec<Modifier> {
    let mut modifiers = vec![];
    for attr in &field.attrs {
        // Nest validified fields
        if attr.path().is_ident(VALIDIFY) {
            modifiers.push(Modifier::Nested);
            continue;
        }

        if !attr.path().is_ident(MODIFY) {
            continue;
        }

        attr.parse_nested_meta(|meta| {
            if meta.path.is_ident(CUSTOM_MODIFIER) {
                let content;
                parenthesized!(content in meta.input);
                let path: syn::Path = content.parse()?;
                modifiers.push(Modifier::Custom { function: path });
                return Ok(());
            }

            if meta.path.is_ident(TRIM_MODIFIER) {
                modifiers.push(Modifier::Trim);
                return Ok(());
            }

            if meta.path.is_ident(LOWERCASE_MODIFIER) {
                modifiers.push(Modifier::Lowercase);
                return Ok(());
            }

            if meta.path.is_ident(UPPERCASE_MODIFIER) {
                modifiers.push(Modifier::Uppercase);
                return Ok(());
            }

            if meta.path.is_ident(CAPITALIZE_MODIFIER) {
                modifiers.push(Modifier::Capitalize);
                return Ok(());
            }

            Err(meta.error("Unrecognized modify parameter"))
        })
        .unwrap_or_else(|e| abort!(e.span(), e));
    }
    modifiers
}
