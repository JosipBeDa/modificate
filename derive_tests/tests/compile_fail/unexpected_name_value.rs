use validify::Validate;

#[derive(Validate)]
struct Email {
    #[validate(not_a = "validator")]
    email: String,
}

fn main() {}
