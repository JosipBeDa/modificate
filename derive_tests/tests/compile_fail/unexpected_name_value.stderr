error: Uncrecognized validate parameter
 --> tests/compile_fail/unexpected_name_value.rs:5:16
  |
5 |     #[validate(not_a = "validator")]
  |                ^^^^^
